javapoet/src/main/java/com/palantir/javapoet/MethodSpec.java

```diff
@@ -143,24 +144,17 @@ void emit(CodeWriter codeWriter, String enclosingName, Set<Modifier> implicitMod
             codeWriter.emit(" ");
         }
 
-        if (isConstructor()) {
-            codeWriter.emit("$L($Z", enclosingName);
+        if (compactConstructor) {
```

Comment:
nit: I'd duplicate the `emitParameters` to make the 3 mutually exclusive cases a bit clearer

```java
if (compactConstructor) {
    codeWriter.emit("$L", enclosingName);
} else if (isConstructor()) {
    codeWriter.emit("$L", enclosingName);
    emitParameters(codeWriter, parameters, varargs);
} else {
    codeWriter.emit("$T $L", returnType, name);
    emitParameters(codeWriter, parameters, varargs);
}
```